barcodes_search/wizards/barcode_search.py

```diff
@@ -0,0 +1,121 @@
+        if len(res) == 1 and not res[0][2]:
+            # Display the result in the according form view
+            record = res[0][1]
+            return record.get_formview_action()
+        else:
+            # Update Wizard Values
+            line_vals = []
+            self.line_ids.unlink()
+            for item in res:
+                line_vals.append((0, 0, {
+                    'field_id': item[0].id,
+                    'item_id': item[1].id,
+                    'item_name': item[1].name,
+                    'extra_data': item[2],
+                }))
+            self.state = 'done'
+            self.line_ids = line_vals
+
+            # Return Action
+            action = self.get_formview_action()
+            action['target'] = 'new'
+            return action
+
+    @api.model
+    def search_by_barcode(self, barcode):
+        """Return a list of tuple (field, record, {extra_data})
+        that matches with the given barcode
+        """
+        barcode_fields = self.get_barcode_fields()
+        res = self._search_by_barcode_barcode_fields(
+            barcode, barcode_fields)
+
+        # Extra search for objects that have a barcode with encoded data
+        # Typical exemple, product with a pattern 20{barcode_base}{NNDDD}
+        # where {NNDDD} is the weight of the product
+        nomenclature_obj = self.env['barcode.nomenclature']
+        nomenclatures = nomenclature_obj.search([])
+        rule_types = self.get_model_by_rule_type()
+        for nomenclature in nomenclatures:
+            parsed_result = nomenclature.parse_barcode(barcode)
+            rule_type = parsed_result['type']
+            if rule_type in rule_types:
+                barcode_fields = self.get_barcode_fields(rule_types[rule_type])
+                base_code = parsed_result['base_code']
+                extra_results = self._search_by_barcode_barcode_fields(
+                    base_code, barcode_fields)
+                for extra_result in extra_results:
+                    extra_result[2]['type'] = rule_type
+                    extra_result[2]['value'] = parsed_result['value']
+                res += extra_results
+        return res
+
+    @api.model
+    def get_barcode_fields(self, model_name=False):
+        """Return a recordset of fields that represent a barcode, in any model.
+        By default, it will return all fields named 'barcode'.
+        Note : Overload that function in a custom module, if you define
+        a barcode field with a name different than 'barcode'"""
+        field_obj = self.env['ir.model.fields']
+        if model_name:
+            domain = [
+                ('name', '=', 'barcode'), ('model', '=', model_name)]
+        else:
+            domain = [
+                ('name', '=', 'barcode'), ('model', '!=', 'barcode.search')]
+        return field_obj.search(domain)
+
+    @api.model
+    def get_model_by_rule_type(self):
+        """Return a dictionary of rule type: model
```

Comment:
Please add a summary line for this method. Multiline docblocks should contain a single line summary, followed by the extended description if necessary. 

I'd make suggestions, but I'm honestly having a bit of trouble understanding what this method is for. Is this to allow you to know what model to search when using a nomenclature of type X (weight, price, ..)?

Reply:
>  Is this to allow you to know what model to search when using a nomenclature of type X (weight, price, ..)?

Yes. 
- I want to search only on certain nomenclature.
- I want to make this list overloadable.

Reply:
Thanks for the elaboration. Does the following hit the intent?

```
Provides a mapping for nomenclature types to Odoo models.

The dictionary returned by this method maps barcode nomenclature rule types (``barcode.rule.type``) to the models that they represent. A barcode nomenclature rule type allows for numerical content to be represented in a barcode by utilizing the `......{NNDDD}` format to indicate the fixed and variable parts of the barcode respectively. 

Note: Modules adding new barcode nomenclature rule types will need to override this method to update the mapping in order to be supported.

For example:
* `weight` is defined in the Odoo `stock` module
* `price` is defined in the Odoo `point_of_sale ` module
* `price_to_weight ` is defined in the OCA `pos_price_to ` module

Returns:
    dict: A mapping of models keyed by their respective barcode rule type (``barcode.rule.type``).
```

That said, aren't we missing a bunch of rule types here?